### PR TITLE
support internlm-xcomposer2-7b & internlm-xcomposer2-4khd-7b

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

related issues: https://github.com/InternLM/lmdeploy/issues/1432 https://github.com/InternLM/lmdeploy/issues/1173

## Modification

Please briefly describe what modification is made in this PR.

## Use cases (Optional)

```python
from lmdeploy import pipeline
from lmdeploy.vl import load_image
# model_id = 'internlm/internlm-xcomposer2-4khd-7b'
model_id = 'internlm/internlm-xcomposer2-7b'
pipe = pipeline(model_id)
im = load_image('../tiger.jpeg')
pipe(('描述这个图片', im))
```

### CODE REVIEW

**Comment by YCG09** [comment]
请问现在internlm-xcomposer2-vl-7b也支持了吗？

File: docs/zh_cn/supported_models/supported_models.md [hunk 1]
```diff
@@ -2,23 +2,24 @@
...
-|   LLaVA(1.5,1.6)   |  7B - 34B  |    Yes    |    -    |    -    |   -   |
-|   InternVL-Chat    |     -      |    Yes    |    -    |    -    |   -   |
+|        模型         |  模型规模   | FP16/BF16 |       KV INT8        | KV INT4 | W4A16 |
+| :-----------------: | :---------: | :-------: | :------------------: | :-----: | :---: |
+|        Llama        |  7B - 65B   |    Yes    |         Yes          |   Yes   |  Yes  |
+|       Llama2        |  7B - 70B   |    Yes    |         Yes          |   Yes   |  Yes  |
+|      InternLM       |  7B - 20B   |    Yes    |         Yes          |   Yes   |  Yes  |
+|      InternLM2      |  7B - 20B   |    Yes    |         Yes          |   Yes   |  Yes  |
+| InternLM-XComposer  |     7B      |    Yes    |          -           |    -    |   -   |
+| InternLM-XComposer2 | 7B, 4khd-7B |    Yes    | -         -        - |         |       |
```
Comment (lvhan028) [inline docs/zh_cn/supported_models/supported_models.md:12]: [InternLM-XComposer, KV INT8] 坐标对应的栏目中，有3个短横线。

File: lmdeploy/archs.py [hunk 2]
```diff
@@ -113,13 +113,19 @@ def autoget_backend_config(
 
 def check_vl_llm(config: dict) -> bool:
     """check if the model is a vl model from model config."""
+    if 'auto_map' in config:
```
Comment (lvhan028) [inline lmdeploy/archs.py:116]: 具体是哪个模型的config没有加 "architectures"呢？

File: lmdeploy/turbomind/supported_models.py [hunk 3]
```diff
@@ -41,6 +43,10 @@ def get_model_arch(model_path: str):
 
     if cfg.get('architectures', None):
         arch = cfg['architectures'][0]
+        if cfg.get('auto_map'):
+            for _, v in cfg['auto_map'].items():
+                if 'InternLMXComposer2ForCausalLM' in v:
+                    arch = 'InternLMXComposer2ForCausalLM'
```
Comment (lvhan028) [inline lmdeploy/turbomind/supported_models.py:49]: arch和能和auto_map不一样么？

File: lmdeploy/vl/model/utils.py [hunk 4]
```diff
@@ -50,6 +51,17 @@ def load_model_from_weight_files(model: nn.Module, folder: str) -> None:
         model.load_state_dict(state_dict, strict=False)
 
 
+@contextmanager
+def add_sys_path(path: Union[str, os.PathLike]) -> Iterator[None]:
```
Comment (lvhan028) [inline lmdeploy/vl/model/utils.py:55]: 加 add_sys_path 的motivation是？

**Comment by irexyc** [comment]
@YCG09 支持的，模型结构，tokenzier看起来都一样。

File: src/turbomind/models/llama/LlamaLinear.h [hunk 5]
```diff
@@ -58,6 +69,41 @@ class LlamaLinear {
...
+
+            invokeMaskAddTwoLinearOutput2(output_data,
+                                          output_data + batch_size * weight.output_dims,
+                                          weight.lora_scale,
+                                          lora_mask,
+                                          batch_size,
+                                          weight.output_dims,
+                                          stream_);
+            sync_check_cuda_error();
+        }
```
Comment (lzhangzz) [inline src/turbomind/models/llama/LlamaLinear.h:106]: This triples the size of all major intermediate buffers. From 600+MB to 2GB for 8K input tokens.

It can be computed as
```
output = lora(x)       // compute lora output first
output = mask(output)
output = x*W + output  // add gemm result to output by gemm API
```
because gemm computes C = alpha\*A\*x + beta*C. This way it will only require `bsz * r` additional workspace and the mask+add kernel can be simplified to mask-only kernel.

File: src/turbomind/models/llama/LlamaDenseWeight.h [hunk 6]
```diff
@@ -57,6 +57,11 @@ struct LlamaDenseWeight {
     size_t     input_dims;
     size_t     output_dims;
     void*      kernel;
+    int        lora_policy;
+    size_t     lora_r;
+    void*      lora_a;
+    void*      lora_b;
+    float      lora_scale;
```
Comment (lzhangzz) [inline src/turbomind/models/llama/LlamaDenseWeight.h:64]: group these using a `struct`

File: src/turbomind/models/llama/LlamaFfnLayer.cc [hunk 7]
```diff
@@ -27,10 +27,15 @@
...
 template<typename T>
-void LlamaFfnLayer<T>::allocateBuffer(size_t token_num)
+void LlamaFfnLayer<T>::allocateBuffer(size_t                     token_num,
+                                      const LlamaDenseWeight<T>* gating,
+                                      const LlamaDenseWeight<T>* inter)
 {
-    inter_buf_          = (T*)allocator_->reMalloc(inter_buf_, sizeof(T) * token_num * inter_size_, false);
-    gating_buf_         = (T*)allocator_->reMalloc(gating_buf_, sizeof(T) * token_num * inter_size_, false);
+    int sz              = sizeof(T) * token_num * inter_size_;
+    int sz_gate         = (gating->lora_policy == 1 && gating->lora_r > 0) ? sz * 3 : sz;
```
Comment (lzhangzz) [inline src/turbomind/models/llama/LlamaFfnLayer.cc:35]: The same logic `lora_policy && lora_r > 0` appears everywhere. Can we simply set `lora_policy = false` when `lora_r == 0`?

File: src/turbomind/models/llama/unified_decoder.cc [hunk 8]
```diff
@@ -183,6 +184,10 @@ void UnifiedDecoder<T>::forward(TensorMap* outputs, const TensorMap* inputs, con
         /// feed-forward network
         TensorMap ffn_inputs{{"ffn_input", {MEMORY_GPU, dtype_, {token_num, hidden_units_}, decoder_output}}};
         TensorMap ffn_outputs{{"ffn_output", {MEMORY_GPU, dtype_, {token_num, hidden_units_}, decoder_output}}};
+        if (lora_params_.policy == 1 && inputs->isExist("lora_mask")) {
```
Comment (lzhangzz) [inline src/turbomind/models/llama/unified_decoder.cc:187]: Why don't foward `lora_mask` when it exists without checking.

File: src/turbomind/models/llama/LlamaV2.cc [hunk 9]
```diff
@@ -131,6 +133,7 @@ void LlamaV2<T>::initialize(const LlamaAttentionParams& attn_params,
                                                  stream_,
                                                  cublas_wrapper_,
                                                  allocator_,
+                                                 lora_params_,
```
Comment (lzhangzz) [inline src/turbomind/models/llama/LlamaV2.cc:136]: `lora_param_` is passed through all abstration layers. And the checking of `lora_param_.policy == 1` does not seem necessary.

File: lmdeploy/turbomind/deploy/source_model/xomposer2.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        """lora config for internlm-xcomposer2-4khd-7b."""
+        num_layer = model_info['num_layer']
+        rank_pattern = []
+        scale_pattern = []
+        rank_qkv = 8
+        rank_wo = 256
+        scale_qkv = 2.0
+        scale_wo = 1.0
+        for i in range(num_layer):
+            for key, rank, scale in zip(['w_qkv', 'wo'], [rank_qkv, rank_wo],
```
Comment (lvhan028) [inline lmdeploy/turbomind/deploy/source_model/xomposer2.py:106]: 看起来每层的attention都是一样的 rank，scale。
定义pattern是为了什么？

File: lmdeploy/turbomind/deploy/source_model/xomposer2.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+                    lora_policy=1,
+                    lora_rank_pattern=rank_pattern,
+                    lora_scale_pattern=scale_pattern)
+
+    def model_info(self):
+        out = super().model_info()
+        params_path = osp.join(self.model_path, 'config.json')
+        with open(params_path) as f:
+            model_arg = json.load(f)
+        if model_arg['max_length'] == 16384:
```
Comment (lvhan028) [inline lmdeploy/turbomind/deploy/source_model/xomposer2.py:123]: 只能用 max_length 来区分么？

File: src/turbomind/models/llama/LlamaDecoderLayerWeight.cc [hunk 12]
```diff
@@ -46,6 +48,36 @@ LlamaDecoderLayerWeight<T>::LlamaDecoderLayerWeight(size_t     head_num,
...
+            std::string full_name = "layers." + std::to_string(layer_idx) + "." + name;
+            if (auto it = lora_params.rank_pattern.find(full_name); it != lora_params.rank_pattern.end()) {
+                rank = it->second;
+                TM_LOG_DEBUG("find rank name=%s, value=%d", full_name.c_str(), rank);
+            }
+            if (auto it = lora_params.scale_pattern.find(full_name); it != lora_params.scale_pattern.end()) {
+                scale = it->second;
+                TM_LOG_DEBUG("find scale name=%s, value=%f", full_name.c_str(), scale);
+            }
+            if (rank && scale > 0) {
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaDecoderLayerWeight.cc:73]: why check this condition?

File: lmdeploy/turbomind/supported_models.py (hunk 3 above)
Comment (irexyc) [reply lmdeploy/turbomind/supported_models.py:51]: image

File: lmdeploy/vl/model/utils.py (hunk 4 above)
Comment (irexyc) [reply lmdeploy/vl/model/utils.py:55]: 为了加载hf model folder里面的函数
```python
from ixc_utils import HD_transform
```

File: src/turbomind/models/llama/LlamaV2.cc [hunk 13]
```diff
@@ -194,8 +211,18 @@ void LlamaV2<T>::updateEmbedding(T* decoder_input, const int bsz, const int* h_i
...
+            if (lora_mask != nullptr) {
+                std::fill_n(mask_ptr + off_dst, (end - begin), 1);
+                *have_embeddings = true;
+            }
         }
         decoder_input += h_input_length[i] * hidden_units_;
+        mask_ptr += h_input_length[i];
+    }
+
+    if (lora_mask != nullptr && *have_embeddings) {
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaV2.cc:223]: 从上面的逻辑看 lora_mask 不为空的时候，have_embeddings 肯定为 true。
所以 have_embeddings 这个参数还有用么？

Comment (irexyc) [reply src/turbomind/models/llama/LlamaV2.cc:223]: plora的话，lora_mask不为空。prefill阶段可能会有embedding，generate阶段不会有embedding。

Comment (lvhan028) [reply src/turbomind/models/llama/LlamaV2.cc:223]: 但是，只要 lora_mask 不为空，bsz 有为空的情况？

File: src/turbomind/models/llama/LlamaLinear.h [hunk 14]
```diff
@@ -60,6 +71,61 @@ class LlamaLinear {
         sync_check_cuda_error();
     }
 
+    void forwardFpLora(T*                         output_data,
+                       const T*                   input_data,
+                       int                        batch_size,
+                       const LlamaDenseWeight<T>& weight,
+                       Type                       type,
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaLinear.h:78]: 参数 type 没有使用到

**Comment by lvhan028** [comment]
test the following cases with `internlm-xcomposer2-4khd-7b`
- [x] converter -> chat (tp 1/2)
- [x] chat with hf (tp 1/2)
- [ ] api_server with hf (tp1/2) + openai api
- [ ] gradio

**Review by lvhan028 — APPROVED** [review]
